### PR TITLE
Add Group Filter for LDAP

### PR DESCRIPTION
**Pull request by johnjoepo**
Adds the ability to define a Group for LDAP Access

### CODE REVIEW

**Comment by alkuzad** [comment]
:+1: this would be the feature that can really differentiate kopf and cerebro.

**Comment by thgrs** [comment]
just as a side note: i just took this diff, applied it to the latest cerebro (0.6.6 right now), rebuilt cerebro and can report, that it works perfectly fine and as expected. any reason why it is not merged into mainline cerebro?

**Comment by barryib** [comment]
Any ETA for this PR?

**Comment by barryib** [comment]
@lmenezes is this feature in the roadmap or there is another way to achieve ldap group filtering? I mean, instead of using nginx or apache in front of cerebro?

**Comment by lmenezes** [comment]
hey all,
sorry for taking so long to get a reply on this.

The reason while this PR hasn't been merged so far is that it *replaces* one authentication method with another(binding to LDAP directly with username+password for binding with a common username + password and searching for the authenticating user), while ideally it should support both. 

There are a few things here which are not really clear to me(I don't know much about LDAP, so perhaps I'm just missing something). I will leave a few comments on the PR and if you are willing to adapt this a little so

File: app/controllers/auth/ldap/LDAPAuthService.scala [hunk 1]
```diff
@@ -1,44 +1,61 @@
...
-  def auth(username: String, password: String): Option[String] = {
-    val env = new Hashtable[String, String](11)
-    env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory")
-    env.put(Context.PROVIDER_URL, s"${config.url}/${config.baseDN}")
-    env.put(Context.SECURITY_AUTHENTICATION, config.method)
-    if (username.endsWith(s"@${config.domain}")) {
-      env.put(Context.SECURITY_PRINCIPAL, username)
-    } else {
-      env.put(Context.SECURITY_PRINCIPAL, s"$username@${config.domain}")
+  def checkGroupMembership(username: String, password: String): Boolean = {
```
[the remark this answers is not in the document]
Comment (johnjoepo) [reply app/controllers/auth/ldap/LDAPAuthService.scala]: You're right. I forgot to implement User Auth check. I will fix this in the next few days.

[the remark this answers is not in the document]
Comment (johnjoepo) [reply app/controllers/auth/ldap/LDAPAuthService.scala]: New commit fixed it

**Comment by Gabisonfire** [comment]
For us the
`groups.contains(config.userGroup)`
always returns false. We've tried debugging and making sure the strings returned match but we ended up replacing with:
`        var isInGroup = false
        groups.foreach(p => if (p == config.userGroup) isInGroup = true)
        isInGroup`

**Comment by FR-TheS** [comment]
Can we have an update about the authentification linked to a specific group?

**Comment by tlied2** [comment]
I have an implementation for this which does not require configured credentials and is based on the existing code. It appears to work well with my OpenLDAP setup, but I do not have AD handy to test (it should work though, as it is a rather generic approach).

Being that we have created a bind/context to test authentication, I just went ahead and reused that context to issue a subtree search (expecting that an authenticated user can perform a search) against a group-dn (like ou=Groups,dc=example,dc=com), not unlike what this PR is proposing, but with a filter. AD is likely to need a different filter definition, but as an example, my filter definition is something like (&(cn=Admins)(uniqueMember=$username), which checks if the user is a member of the group named Admins. If I get 1 (and only 1; trying to guard against misconfig) result the user is authorized.

With a little more code it should be easy enough to add a search which validates an attribute set on a user entry rather than doing a subtree search if anyone would find it helpful. For example, I have a different objectClass set for my active users than my inactive users.

I am happy to open a PR with what I have at present (which appears to work for me), and I'll have a better look at how AD arranges groups, but I wanted to solicit feedback before I got too carried away.

**Comment by moliware** [comment]
Hi! I based on this PR and @tlied2's feedback I added #307 . I would be awesome if I could receive your feedback.

Thank you!

**Comment by Chrobak77** [comment]
> Hi! Based on this PR and @tlied2's feedback I added #307 . I would be awesome if I could receive your feedback.
> 
> Thank you!

Hello,
we are currently testing and it seems to be working, however only partialy.
Authentication works after setup, but there are some issues.

It seems to me that user-format parameter is ignoring whatever you put there.
Also it would be nice to have proper readme with examples as it is a bit confusing ,for example setting the user-template (which is required even though it is not used in the and at all).

Could you maybe provide logback.xml classes in order to setup maximum debugging of LDAP search a login?


auth = {
  type: ldap
  settings {
    url = "ldaps://ourldap.com:3269"
    base-dn = "OU=Users,DC=our,DC=domain,DC=com"
    method = "simple"
    user-format="%s@domain.com"
    user-template = "CN={0},DC=our,DC=domain,DC=com"
    bind-dn = "CN=ldapacc,OU=application,OU=ServiceAccounts,DC=domain,DC=com"
    bind-pw = "xxxxx"
    group-search {
      base-dn = "OU=Users,DC=our,DC=domain,DC=com"
      user-attr = "UserPrincipalName"
      group = "memberOf=CN=SuperGroup,OU=AppAccess,OU=Groups,DC=domain,DC=com"
    }

Thanks

**Comment by Scoobed** [comment]
I was reviewing this code.  BindDN before was more flexible before hardcoding this bindUser@user.domain  props.put(Context.SECURITY_PRINCIPAL, config.bindDN + "@" + config.userDomain) in the latest commit.  Is it possible to revert this and just go back to bindDN?

**Comment by moliware** [comment]
Done in #307 and released in 0.8.2

I tested it in AD and works fine so hopefully it works for you too. See doc here : https://github.com/lmenezes/cerebro#ldap-config

Thanks!

**Pull request closed without merging**